samples/api/swapchain_recreation/swapchain_recreation.cpp

```diff
@@ -0,0 +1,950 @@
+	std::vector<PresentOperation> history_to_keep;
+	while (!present_history.empty())
+	{
+		PresentOperation &present_operation = present_history.back();
+
+		// If this is about an older swapchain, let it be.
+		if (present_operation.image_index == INVALID_IMAGE_INDEX)
+		{
+			assert(present_operation.cleanup_fence != VK_NULL_HANDLE);
+			break;
+		}
+
+		// Reset the index, so it's not processed in the future.
+		present_operation.image_index = INVALID_IMAGE_INDEX;
+
+		if (present_operation.cleanup_fence != VK_NULL_HANDLE)
+		{
+			// If there is already a fence associated with it, let it be cleaned up once
+			// the fence is signaled.
+			history_to_keep.push_back(std::move(present_operation));
+		}
+		else
+		{
+			assert(present_operation.present_semaphore != VK_NULL_HANDLE);
+
+			// Otherwise accumulate it in cleanup data.
+			cleanup.semaphores.push_back(present_operation.present_semaphore);
+
+			// Accumulate any previous swapchains that are pending destruction too.
+			for (SwapchainCleanupData &swapchain : present_operation.old_swapchains)
+			{
+				old_swapchains.emplace_back(swapchain);
+			}
+			present_operation.old_swapchains.clear();
+		}
+
+		present_history.pop_back();
+	}
+	std::move(history_to_keep.begin(), history_to_keep.end(), std::back_inserter(present_history));
+
+	if (cleanup.swapchain != VK_NULL_HANDLE || !cleanup.semaphores.empty())
+	{
+		old_swapchains.emplace_back(std::move(cleanup));
+	}
+}
+
+VkSemaphore SwapchainRecreation::get_semaphore()
+{
+	// If there is a free semaphore, return it
+	if (!semaphore_pool.empty())
+	{
+		VkSemaphore semaphore = semaphore_pool.back();
+		semaphore_pool.pop_back();
+		return semaphore;
+	}
+
+	VkSemaphore           semaphore = VK_NULL_HANDLE;
+	VkSemaphoreCreateInfo create_info{VK_STRUCTURE_TYPE_SEMAPHORE_CREATE_INFO};
+
+	VK_CHECK(vkCreateSemaphore(get_device_handle(), &create_info, nullptr, &semaphore));
+
+	return semaphore;
+}
+
+void SwapchainRecreation::recycle_semaphore(VkSemaphore semaphore)
+{
+	semaphore_pool.push_back(semaphore);
+}
+
+VkFence SwapchainRecreation::get_fence()
+{
+	// If there is a free fence, return it
+	if (!fence_pool.empty())
+	{
+		VkFence fence = fence_pool.back();
+		fence_pool.pop_back();
+		return fence;
+	}
+
+	VkFence           fence = VK_NULL_HANDLE;
+	VkFenceCreateInfo create_info{VK_STRUCTURE_TYPE_FENCE_CREATE_INFO};
+
+	VK_CHECK(vkCreateFence(get_device_handle(), &create_info, nullptr, &fence));
+
+	return fence;
+}
+
+void SwapchainRecreation::recycle_fence(VkFence fence)
+{
+	fence_pool.push_back(fence);
+
+	VK_CHECK(vkResetFences(get_device_handle(), 1, &fence));
+}
+
+VkPhysicalDevice SwapchainRecreation::get_gpu_handle()
+{
+	return get_device().get_gpu().get_handle();
+}
+
+VkDevice SwapchainRecreation::get_device_handle()
+{
+	return get_device().get_handle();
+}
+
+SwapchainRecreation::SwapchainRecreation()
+{
+	LOGI("USAGE:");
+	LOGI(" - Press v to enable v-sync");
+	LOGI(" - Press n to disable v-sync");
```

Comment:
Maybe move this output at some place such that it's more visible to the user. Like in `SwapchainRecreation::prepare` (after calling `VulkanSample::prepare`)

Reply:
Sounds good, done.